app/controllers/extra_locales_controller.rb

```diff
@@ -11,67 +11,75 @@ class ExtraLocalesController < ApplicationController
-        hash = ExtraLocalesController.bundle_js_hash(bundle)
-        immutable_for(1.year) if hash == version
+    def bundle_js_hash(bundle)
+      bundle_key = "#{bundle}_#{I18n.locale}"
+      if bundle.in?(BUNDLES)
+        site = RailsMultisite::ConnectionManagement.current_db
+
+        js_digests[site] ||= {}
+        js_digests[site][bundle_key] ||= begin
+          js = bundle_js(bundle)
+          js.present? ? Digest::MD5.hexdigest(js) : nil
+        end
       else
-        raise Discourse::InvalidParameters.new(:v)
+        js_digests[bundle_key] ||= Digest::MD5.hexdigest(bundle_js(bundle))
       end
     end
 
-    render plain: ExtraLocalesController.bundle_js(bundle), content_type: "application/javascript"
-  end
+    def url(bundle)
+      "#{Discourse.base_path}/extra-locales/#{bundle}?v=#{bundle_js_hash(bundle)}"
+    end
 
-  def self.bundle_js_hash(bundle)
-    if bundle == OVERRIDES_BUNDLE
-      site = RailsMultisite::ConnectionManagement.current_db
+    def client_overrides_exist?
+      bundle_js_hash(OVERRIDES_BUNDLE).present?
+    end
+
+    def bundle_js(bundle)
+      locale_str = I18n.locale.to_s
+      bundle_str = "#{bundle}_js"
 
-      @by_site ||= {}
-      @by_site[site] ||= {}
-      @by_site[site][I18n.locale] ||= begin
-        js = bundle_js(bundle)
-        js.present? ? Digest::MD5.hexdigest(js) : nil
+      case bundle
+      when OVERRIDES_BUNDLE
+        JsLocaleHelper.output_client_overrides(locale_str)
+      when MF_BUNDLE
+        JsLocaleHelper.output_MF(locale_str)
+      else
+        JsLocaleHelper.output_extra_locales(bundle_str, locale_str)
       end
-    else
-      @bundle_js_hash ||= {}
-      @bundle_js_hash["#{bundle}_#{I18n.locale}"] ||= Digest::MD5.hexdigest(bundle_js(bundle))
     end
-  end
 
-  def self.url(bundle)
-    "#{Discourse.base_path}/extra-locales/#{bundle}?v=#{bundle_js_hash(bundle)}"
+    def clear_cache!
+      site = RailsMultisite::ConnectionManagement.current_db
+      js_digests.delete(site)
+    end
   end
 
-  def self.client_overrides_exist?
-    bundle_js_hash(OVERRIDES_BUNDLE).present?
-  end
+  def show
+    bundle = params[:bundle]
+    raise Discourse::InvalidAccess.new if !valid_bundle?(bundle)
 
-  def self.bundle_js(bundle)
-    locale_str = I18n.locale.to_s
-    bundle_str = "#{bundle}_js"
+    version = params[:v]
+    if version.present?
+      raise Discourse::InvalidParameters.new(:v) unless version.to_s.size == MD5_HASH_LENGTH
 
-    if bundle == OVERRIDES_BUNDLE
-      JsLocaleHelper.output_client_overrides(locale_str)
-    else
-      JsLocaleHelper.output_extra_locales(bundle_str, locale_str)
+      hash = ExtraLocalesController.bundle_js_hash(bundle)
+      immutable_for(1.year) if hash == version
     end
-  end
 
-  def self.clear_cache!
-    site = RailsMultisite::ConnectionManagement.current_db
-    @by_site&.delete(site)
+    render plain: ExtraLocalesController.bundle_js(bundle), content_type: "application/javascript"
```

Comment:
It looks like both `.bundle_js_hash` and `.bundle_js` cause the full JS to be compiled. Seems a little wasteful to do it twice for every request. Could we refactor slightly to improve that?

Maybe something along the lines of:

```ruby
version = params[:v]
if version.present?
  raise Discourse::InvalidParameters.new(:v) unless version.to_s.size == MD5_HASH_LENGTH
end
content, hash = ExtraLocalesController.bundle_js(bundle)
immutable_for(1.year) if hash == version

render plain: content, content_type: "application/javascript"
```

(I realise this problem existed before this PR. But it would be nice to improve things while we're here)

Reply:
Done! I made minimal changes in the end. I still think we should really refactor the whole JS bundle concept to a proper abstraction if we revisit this controller, for example when we’ll serve the classic locales with it.